File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../../input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

// 정점의 개수 N, 간선의 개수 M, 탐색 시작 정점 V
const [n, m, v] = input[0].split(' ').map(Number);
const points = input.slice(1).map((x) => x.split(' ').map(Number));

path();

function path() {
    // 그래프 초기화
    const graph = Array.from({ length: n + 1 }, () => []);
    points.forEach(([start, end]) => {
        graph[start].push(end);
        graph[end].push(start);
    });

    // 정점 번호가 작은 것을 먼저 방문하기 위해 정렬
    graph.forEach((x) => x.sort((a, b) => a - b));

    dfs(v);
    bfs(v);
    
    function dfs(start) {
        const visited = Array.from({ length: n + 1 }).fill(false);
        const action = [];
        const stack = [start];

        while (stack.length > 0) {
            const curNode = stack.pop();
            if (!visited[curNode]) {
                visited[curNode] = true;
                action.push(curNode);

                // 스택에 추가 시, 내림차순 정렬하여 작은 정점이 나중에 push되게 해야
                // pop할 때 작은 정점부터 방문할 수 있다.
                const nextNodes = graph[curNode].slice().sort((a, b) => b - a);
                stack.push(...nextNodes);
            }
        }
        console.log(action.join(' '));
    }

		function bfs(start) {
		    const visited = Array(n + 1).fill(false);
		    const action = [];
		    const queue = [start];
		    visited[start] = true; 
		
		    while (queue.length > 0) {
		        const curNode = queue.shift();
		        action.push(curNode);  
		
		        graph[curNode].forEach(point => {
		            if (!visited[point]) {
		                visited[point] = true;  
		                queue.push(point);
		            }
		        });
		    }
		    console.log(action.join(' '));
		}
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

lines = sys.stdin.read().splitlines()
n, m, v = map(int, lines[0].split())
connections = lines[1:1+m]
graph = list([] for i in range(0, n + 1))

# 그래프 만들기
for connection in connections:
    a, b = map(int, connection.split())
    graph[a].append(b)
    graph[b].append(a)

def dfs(start): # stack
    visited = []
    stack = []

    stack.append(start)
    
    while stack:
        node = stack.pop()

        # stack은 반대로 정렬(내림차순)을 해서 넣어줘야 된다.
        # 뒤에가 먼저 빠지기 때문에!
        if node not in visited:
            visited.append(node)
            sortedNodes = sorted(graph[node], reverse=True)
            stack.extend(sortedNodes)    
            
    return ' '.join(map(str, visited))

def bfs(start): # queue
    visited = []
    queue = deque()

    queue.append(start)
    
    while queue:
        node = queue.popleft()
        
        if node not in visited:
            visited.append(node)
            sortedNodes = sorted(graph[node])
            queue.extend(sortedNodes)

    return ' '.join(map(str, visited))

print(dfs(v))
print(bfs(v))
